### ISSUE TITLE
Provide concise API to prevent extraneous api calls with SSR (withPageAuth)

### ISSUE THREAD

**Issue by Izhaki** [auth0/nextjs-auth0#898]
### Checklist

- [X] I have looked into the [README](https://github.com/auth0/nextjs-auth0#readme) and have not found a suitable solution or answer.
- [X] I have looked into the [examples](https://github.com/auth0/nextjs-auth0/blob/main/EXAMPLES.md) and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://auth0.github.io/nextjs-auth0/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/nextjs-auth0/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com/c/sdks/5) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

I would like to prevent extraneous API calls with non-auth-protected pages, when using SSR.

These extra calls are redundant in SSR settings and:
- Can be costly with high user volume systems.
- Reduce time to final render due to loading state and request roundtrip.

### Describe the ideal solution

Provide a `withPageAuth` helper, similar to [WithPageAuthRequired](https://auth0.github.io/nextjs-auth0/modules/helpers_with_page_auth_required.html#withpageauthrequired), but that can be used for non protected pages.

```js
export const getServerSideProps = withPageAuth();
```

Potentially, the future API can deprecate `WithPageAuthRequired` and provide a `required` property to the helper above.

```js
export const getServerSideProps = withPageAuth({ required: true});
```

#### useUser()

Currently if you provide `undefined` as the `user` prop to the `UserProvider`, it will try to fetch from `/me`. Ideally there would be a way disable this behaviour in cases where the user is not logged in, yet the information is arriving from the server.

One option is to distinguish between `undefined` and `null`. But more explicitly we could:

```jsx
export default function App({ Component, pageProps }) {
  const { user } = pageProps;

  return (
    <UserProvider user={user} disableProfileFetch>
      <Component {...pageProps} />
    </UserProvider>
  );
}
```

### Alternatives and current workarounds

Current alternative is to DIY using [getServerSidePropsWrapper](https://auth0.github.io/nextjs-auth0/modules/helpers_get_server_side_props_wrapper.html) and `getSession` to achieve the same thing.

To prevent fetching in case of logout, the server can return `user: {}` and client code could check if the property `name` exists on that object (if not, the user is not logged in).

### Additional context

I am trying this lib for the first time.

Was a bit surprised to see a loading state with SSR, and a failed network request to `auth/me` when the user is logged out.

I would assume with SSR rendering the server will pass the required information to the client, rather the client making an API call.

An investigation showed that it is possible, although with different API for protected and non-protected pages.

**Comment by adamjmcgrath** [comment]
Hi @Izhaki - thanks for raising this

> Current alternative is to DIY using [getServerSidePropsWrapper](https://auth0.github.io/nextjs-auth0/modules/helpers_get_server_side_props_wrapper.html) and getSession to achieve the same thing.

You no longer need `getServerSidePropsWrapper` in the new beta, see the [migration guide](https://github.com/auth0/nextjs-auth0/blob/v2.0.0-beta.2/V2_MIGRATION_GUIDE.md#getserversidepropswrapper-has-been-removed) for more info

> Currently if you provide undefined as the user prop to the UserProvider, it will try to fetch from /me. Ideally there would be a way disable this behaviour in cases where the user is not logged in, yet the information is arriving from the server.

We looked into this issue when investigating the Beta and decided to leave the behaviour as is as disabling the check didn't work well with CSR user checks like the CSR `withPageAuthRequired` - if you want the longer details , here's the outcome from our investigation:

> We decided to drop it because it results in a redirect loop for CSR pages, when checking the session is disabled and the user navigates to a CSR page right after logging in.
> At that point, the user is logged in but the frontend doesn't know it yet. Thus withAuthRequired tries to redirect to the login page, which in turn redirects to Auth0, which promptly redirects back because the user is in fact logged in but the frontend still doesn't know about it, looping.
> It only works correctly when a SSR page is visited first, right after login, because the user value gets filled in the React context. After that, when navigating to a CSR page it just gets returned.
> Unfortunately the new prop is set at the app-level –it wraps up the App component. There's no way to only have it be false for SSR pages because at this point, in the App component, pageProps.user is undefined.

If you're not using `withPageAuthRequired` CSR you can workaround this by selectively rendering the UserProvider or overriding the userFetcher

Closing, feel free to ping me if you have any other questions

**Comment by Izhaki** [comment]
I've like to loop back on this, as I still believe there is scope for potential improvements.

> We decided to drop it because it results in a redirect loop for CSR pages, when checking the session is disabled and the user navigates to a CSR page right after logging in.
At that point, the user is logged in but the frontend doesn't know it yet. Thus withAuthRequired tries to redirect to the login page, which in turn redirects to Auth0, which promptly redirects back because the user is in fact logged in but the frontend still doesn't know about it, looping.

I guess `withAuthRequired` means the one from `/client`, not the server version as the server version will not be used with CSR.

To further illustrate, consider:

```js
export default function App({ Component, pageProps }) {
  const { user, isSsr } = pageProps;

  return (
    <UserProvider user={user} disableProfileFetch={isSsr}>
      <Component {...pageProps} />
    </UserProvider>
  );
}
```

Basically, if we are doing SSR (determined by the `isSsr` flag), then at this point we are certain about the user session, so no real need to hit `auth/me`.

CSR pages will not have `isSsr` and hence their behaviour will not be affected.

Hope this makes sense.
